Polynomial control

andreadelprete opened the pull request:
This PR introduces the possibility to use an arbitrary discretization for the control inside the time step. Thanks to this, we will no longer be limited to piece-wise constant control inputs, but could use (e.g.) polynomials. I've already implemented first and second-order polynomials, but higher orders could also be used. 

I've introduced a new abstract class `IntegratedActionModelAbstract`, which inherits from `ActionModelAbstract`. When creating an `IntegratedActionModelAbstract`, a `ControlAbstract` object needs to be specified to handle the control discretization. The different integration schemes (Euler and RK4 at the moment) then inherits from `IntegratedActionModelAbstract`.

proyan commented:
Give us some time for reviewing Andrea, this is a big PR

andreadelprete commented:
> Give us some time for reviewing Andrea, this is a big PR

Sure, I agree it's big, it took me a month of work! :)

proyan commented:
Thanks Andrea for this PR.

I'm okay with the proposal, and the polynomial logic that you introduced. 
However, I have three main concerns about the API that you used in the control-base class:
1) We've tried to keep the structure `model`, `data`, `calc`, `calcdiff` throughout the repo. It would be better if we follow the same structure for the control-base classes as well. 
Thus, Most of the temporary variables could be pre-defined in the ControlDataAbstract, and the API would change:
instead of 
```c++
virtual void value(double t, const Eigen::Ref<const VectorXs>& p, Eigen::Ref<VectorXs> u_out) const = 0;
```
we would have 
```c++
virtual void calc(data, t, p);
```
In this case, the output `u_out` is stored in data, and thus the distinction between `value` and `value_u` is no longer required.
Similarly, `dValue` would become `calcDiff`, and dValue and dValue_J would merge.

2) The definition of control parameters `p` and control `u` seems to be changing. The control `u` in the Integrated ActionModel is the parameter `p` in the ControlBase, and then we have the additional `u_diff` which is the actual control at a time step. Perhaps parameter is not the best name for the control base? I think `u_poly` and and `u_t` might be more descriptive?

3) There are multiple places where Temporary vectors are being defined, for example all functions of the control-base class, in `IntegratedActionModelEulerTpl::quasiStatic` etc. I haven't marked all of them. Most of these could be removed with the definition of the ControlData class, and others we can deal with separately.

andreadelprete commented:
> Thanks Andrea for this PR.

Thanks for the review!

>     1. We've tried to keep the structure `model`, `data`, `calc`, `calcdiff` throughout the repo. It would be better if we follow the same structure for the control-base classes as well.
>        Thus, Most of the temporary variables could be pre-defined in the ControlDataAbstract, and the API would change:

I am fine with this change, even though I am afraid that using the same name of the methods of ActionModel could lead to confusion for the users.

> the distinction between `value` and `value_u` is no longer required.

I had introduced `value_u` and `dValue_J` to use them in the Python bindings, following the example of the bindings of the State class.

>     1. The definition of control parameters `p` and control `u` seems to be changing. The control `u` in the Integrated ActionModel is the parameter `p` in the ControlBase, and then we have the additional `u_diff` which is the actual control at a time step. Perhaps parameter is not the best name for the control base? I think `u_poly` and and `u_t` might be more descriptive?

My code went through a big refactoring regarding the names of `u` and `p`, which is why it may not be super clean. I find `u_poly` a bit limited in scope, because we are not constrained to polynomial parameterizations of the control. I agree that `p` is not a great name. The key problem is that the `u` given as input to the `IntegratedActionModel` is not the same `u` given as input to the `DifferentialActionModel`, so there must be a change of name somewhere. I thought that `u_diff` was clear because it refers to the `u` of the `DifferentialActionModel`. Maybe for the parameters we could use `u_params`? But be aware that this `u_params` (or whatever we end up calling it) is gonna be called `u` in `IntegratedActionModel::calc`, because from the outside it is seen as the control input of the discrete-time action model.
 
>     2. There are multiple places where Temporary vectors are being defined, for example all functions of the control-base class, in `IntegratedActionModelEulerTpl::quasiStatic` etc. I haven't marked all of them. Most of these could be removed with the definition of the ControlData class, and others we can deal with separately.

Sorry about that, to be honest I hadn't paid much attention to the temporaries. I'm gonna take care of that.

proyan commented:
> I am fine with this change, even though I am afraid that using the same name of the methods of ActionModel could lead to confusion for the users.

Since almost all classes follow the same Model Data structure, I don't think it would be that confusing. It might be useful to mention somewhere in the naming that what we are dealing with is a polynomial representation of the control, and not the discrete constant control that is always assumed. 
Perhaps if we rename the class to `PolynomialControlModelAbstract` and `PolynomialControlDataAbstract`, that might help. What do you think?

andreadelprete commented:
> > I am fine with this change, even though I am afraid that using the same name of the methods of ActionModel could lead to confusion for the users.
> 
> Since almost all classes follow the same Model Data structure, I don't think it would be that confusing. It might be useful to mention somewhere in the naming that what we are dealing with is a polynomial representation of the control, and not the discrete constant control that is always assumed.
> Perhaps if we rename the class to `PolynomialControlModelAbstract` and `PolynomialControlDataAbstract`, that might help. What do you think?

Are you sure we wanna constrain ourselves to polynomials? Other parameterizations may be used in the future.
If `ControlModelAbstract` is too generic we might use `ControlDiscretizationModelAbstract` or `ControlParameterizationModelAbstract`.

proyan commented:
You are right, it is better as ControlModelAbstract. The Polynomial classes are inheriting from this base class, so the polynomial is already shown in the names

andreadelprete commented:
Ok then. What about the names `u_params` and `u_diff` for the control parameters and the instantaneous control? Otherwise we could also use simply `u` for the parameters and `tau` for the instantaneous control.

proyan commented:
I don't think we can use tau for the input to DifferentialModel. There are two levels here that I see

ControlModel : action input -> differential input
ActuationModel : differential input -> pinocchio input

For the pinocchio input, we are already using `tau`.
This is going to be confusing for the new users, but I guess u_params and u_diff is the best choice we have here.

andreadelprete commented:
Agreed! The only disadvantage I see in using such verbose names is that the Jacobian matrices will have rather unfriendly names, such as `ddli_dudiffduparams`. If we went for shorter names, such as `u` and `p`, the same Jacobian would simply be `ddli_dudp`. Do you see any solution to this? For instance, would it be acceptable to use the longer names (u_diff, u_params) for the quantities themselves, but use shorter 1-letter names (u, p) when these variables are the argument of differentiation? Of course this should be properly documented in the code!

proyan commented:
Yes, let's do that.
In the python bindings, in order not to break the code of people using named arguments for calc(data, x, u), let's keep the name change only in the documentation. So for action models, the documentation would mention u_params and for the diff action model, the documentation would mention u_diff

On June 21, 2021 8:11:51 AM GMT+02:00, Andrea Del Prete ***@***.***> wrote:
>Agreed! The only disadvantage I see in using such verbose names is that
>the Jacobian matrices will have rather unfriendly names, such as
>`ddli_dudiffduparams`. If we went for shorter names, such as `u` and
>`p`, the same Jacobian would simply be `ddli_dudp`. Do you see any
>solution to this? For instance, would it be acceptable to use the
>longer names (u_diff, u_params) for the quantities themselves, but use
>shorter 1-letter names (u, p) when these variables are the argument of
>differentiation? Of course this should be properly documented in the
>code!
>
>-- 
>You are receiving this because you commented.
>Reply to this email directly or view it on GitHub:
>https://github.com/loco-3d/crocoddyl/pull/960#issuecomment-864756617

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

cmastalli commented:
>  If we went for shorter names, such as `u` and `p`, the same Jacobian would simply be `ddli_dudp`

I also like this proposition: `u` and `p` are fine. And if I understood correctly, you're proposing to discard the `u_diff` and `u_params` alternatives, right? If so, I also agree with that. Note that I would wish to have `u` and `p` for Python API, and `u_diff` and `u_params` for c++ API.

> `ControlParameterizationModelAbstract`.

I like more this option, which it is also linked with the `p` notation.

cmastalli commented on include/crocoddyl/core/integrator/rk4.hpp:
```diff
@@ -123,24 +136,32 @@ struct IntegratedActionDataRK4Tpl : public ActionDataAbstractTpl<_Scalar> {
   virtual ~IntegratedActionDataRK4Tpl() {}
 
   VectorXs dx;
+  std::vector<VectorXs> u_diff;
```
I believe it is better to use `std::array` as we are supporting c++11>. But, this might need to include extra code for binding them (similar to this code: https://github.com/loco-3d/crocoddyl/blob/devel/bindings/python/crocoddyl/utils/vector-converter.hpp#L117).

Feel free to ignore this request if you don't how to handle it.

andreadelprete replied:
Since I am short on time I prefer to skip this as well.

cmastalli replied:
I would like to give a try on this topic (and in this PR) as it leads to an improvement in the code.
However, this will not block the merging of this PR. In worse case scenario, I will create an issue to track this needed change.

Please let me reopen this conversation (only for my own purpose)

cmastalli commented on include/crocoddyl/core/integrator/rk4.hxx:
```diff
@@ -11,44 +11,62 @@
 #include "crocoddyl/core/utils/exception.hpp"
 #include "crocoddyl/core/integrator/rk4.hpp"
 
+using namespace std;
```
Please do not include this line as it might introduce bugs. 
Instead,  use the `std` namespace in the respective code.

cmastalli commented on unittest/factory/control.hpp:
```diff
@@ -0,0 +1,55 @@
+#include <boost/shared_ptr.hpp>
+#include "crocoddyl/core/fwd.hpp"
+#include "crocoddyl/core/control-base.hpp"
+// #include "crocoddyl/core/numdiff/control.hpp"
```
Remove this commented line

cmastalli commented on unittest/test_controls.cpp:
```diff
@@ -0,0 +1,101 @@
+
+//----------------------------------------------------------------------------//
+
+void test_Dvalue_num_diff(ControlTypes::Type control_type) {
```
Rename the `test_` functions with the agreed names discussed in the Control API.

cmastalli reviewed (changes requested):
Thank you @andreadelprete for this effort. The code is quite nice, nonetheless it requires an extra effort.

On top of @proyan comments, I have left extra points regarding the control API. Below, I summarize these points:
  1. Remove `resize` and also deprecate `set_differential` (only method that uses `resize`)
  2. Rename `multiplyByDValue` and `multiplyByDValueTransposeBy` to `multiplyByJacoban` and `Rename function to multiplyJacobianTransposeBy`, respectively. It follows up the suggestion of renaming `value` and `dValue` by `calc` and `calcDiff`.
  3. Rename `value_inv` to `params` to follow convention used in state classes.
  4. Remove `ControlAbstractTpl::ControlAbstractTpl` constructor (also for the various implementations)

Note that I have include a bit high-level comments (to avoid some many of them). However, I am keen to review it again after you have resolved each point (mine and @proyan ones)

cmastalli commented:
@andreadelprete -- follow this [instructions](https://github.com/loco-3d/crocoddyl/wiki/Developer-Guide) to quickly fix the code-format error.

andreadelprete commented:
> > If we went for shorter names, such as `u` and `p`, the same Jacobian would simply be `ddli_dudp`
> 
> I also like this proposition: `u` and `p` are fine. And if I understood correctly, you're proposing to discard the `u_diff` and `u_params` alternatives, right? If so, I also agree with that. Note that I would wish to use `u` and `p` for both Python and c++ APIs.

No, I wasn't suggesting to discard `u_diff` and `u_params` but to use these long names for the variables containing the control and the control parameters, while using the short names (e.g., `u` and `p`) to indicate differentiation w.r.t. these variables, just to avoid long unreadable names. However, for coherence with previous code and with the DDP notation, I think the short name `u` should be used for the "control action input", which corresponds to `u_params`. Therefore we need to find a short name for `u_diff`, which can neither be `u` nor `tau`. Any idea? Maybe `w`? Or `z`? Or `y`?

> > `ControlParameterizationModelAbstract`.
> 
> I like more this option, which it is also linked with the `p` notation.

Yes, I also find this more explicit so we could go for this one.

andreadelprete commented on bindings/python/crocoddyl/core/integrator/euler.cpp:
```diff
@@ -1,13 +1,14 @@
 
 #include "python/crocoddyl/core/core.hpp"
-#include "python/crocoddyl/core/action-base.hpp"
+#include "python/crocoddyl/core/integr-action-base.hpp"
```
`int` made me think of `integer`, that's why I went for the longer `integr`, but I am fine with this.

cmastalli replied:
Yes, you're right! `init` is not the best option here. According to this [dictionary](https://acronyms.thefreedictionary.com/integration), the right acronym is `integ`.

Would you agree with this proposition?

andreadelprete commented:
>     4. Remove `ControlAbstractTpl::ControlAbstractTpl` constructor (also for the various implementations)

I don't get why I should remove the constructor of `ControlAbstractTpl`. Since ControlAbstractTpl has the member variables `np` and `nu` that are inherited by all the child classes, it seems appropriate that the values of these variables are initialized by the constructor of ControlAbstractTpl, doesn't it?

proyan commented:
> No, I wasn't suggesting to discard u_diff and u_params but to use these long names for the variables containing the control and the control parameters, while using the short names (e.g., u and p) to indicate differentiation w.r.t. these variables, just to avoid long unreadable names. However, for coherence with previous code and with the DDP notation, I think the short name u should be used for the "control action input", which corresponds to u_params. Therefore we need to find a short name for u_diff, which can neither be u nor tau. Any idea? Maybe w? Or z? Or y?

Don't really have a preference for the `u_diff`. Perhaps `ud`? You could take your pick with what you feel suitable.

cmastalli commented:
> However, for coherence with previous code and with the DDP notation, I think the short name `u` should be used for the "control action input", which corresponds to `u_params`. Therefore we need to find a short name for `u_diff`, which can neither be `u` nor `tau`. Any idea? Maybe `w`? Or `z`? Or `y`?

I would go for `w`, which is the letter that comes after `u`.

> I don't get why I should remove the constructor of `ControlAbstractTpl`. Since ControlAbstractTpl has the member variables `np` and `nu` that are inherited by all the child classes, it seems appropriate that the values of these variables are initialized by the constructor of ControlAbstractTpl, doesn't it?

I refereed to this constructor: https://github.com/andreadelprete/crocoddyl/blob/polynomial-control/include/crocoddyl/core/control-base.hpp#L49

> Don't really have a preference for the `u_diff`. Perhaps `ud`? You could take your pick with what you feel suitable.

I would prefer to go for one letter. I think `w` is the best choice here, but also `z` sounds well.

andreadelprete commented on include/crocoddyl/core/control-base.hpp:
```diff
@@ -0,0 +1,150 @@
+  ControlAbstractTpl();
+  virtual ~ControlAbstractTpl();
+
+  virtual void resize(const std::size_t nu) = 0;
```
I just realized why I had introduced this. It's because of the method `set_differential` of the `IntegratedActionModels`. When you set a new differential with that method you might need to resize the control model. Therefore I see two options: either I remove both these methods (IntegratedActionModels::set_differential and Control::resize) or I keep them both. What do you prefer?

cmastalli replied:
Thanks for the explanation.
Yes, I like the idea to remove 'set_differential' and this constructor.

andreadelprete commented on include/crocoddyl/core/control-base.hpp:
```diff
@@ -0,0 +1,150 @@
+   * @param[out] p_lb   Control parameters lower bound
+   * @param[out] p_ub   Control parameters upper bound
+   */
+  virtual void convert_bounds(const Eigen::Ref<const VectorXs>& u_lb, const Eigen::Ref<const VectorXs>& u_ub,
```
I don't see the reason for storing the bounds here. The bounds are already stored in the ActionModel, so we would be duplicating the information. I think the ControlParametrizationModel should simply provide a method to convert the bounds from the differential control space to the parameter space, but the values of the bounds should be stored elsewhere (as it currently is). 
Moreover, adding the bounds in the constructor seems like asking too much information to build a ControlParametrizationModel because in the part of the code where the control parametrization is created the information on the control bounds may not be available. Therefore we should at least provide a version of the constructor without bounds. This means we should have also the `set_lb` and `set_ub` methods, and we would end up replacing the current one-line instruction in `IntegratedActionModelAbstractTpl<Scalar>::init()`:
```
control_->convert_bounds(differential_->get_u_lb(), differential_->get_u_ub(), p_lb, p_ub);
```
with 4 lines of code:
```
control_->set_lb(differential_->get_u_lb());
control_->set_ub(differential_->get_u_ub());
p_lb = control_->get_lb();
p_ub = control_->get_ub();
```

cmastalli replied:
> I don't see the reason for storing the bounds here. The bounds are already stored in the ActionModel, so we would be duplicating the information. I think the ControlParametrizationModel should simply provide a method to convert the bounds from the differential control space to the parameter space, but the values of the bounds should be stored elsewhere (as it currently is).

Fair enough! Let's don't do this (as you suggested)!

andreadelprete commented on include/crocoddyl/core/control-base.hpp:
```diff
@@ -0,0 +1,150 @@
+  virtual MatrixXs dValue_J(double t, const Eigen::Ref<const VectorXs>& p) const;
+
+  /**
+   * @brief Compute the product between a specified matrix and the Jacobian of the control (with respect to the parameters)
```
"The Jacobian of the control parameters" does not mean what I meant. What I meant is "the Jacobian of the control with respect to the parameters". The control is a function of the (control) parameters, so I am talking about the derivative of this function (the control) with respect to its input (the parameters).

andreadelprete commented on include/crocoddyl/core/control-base.hpp:
```diff
@@ -0,0 +1,150 @@
+   * @param[in]  A      A matrix to multiply times the Jacobian
+   * @param[out] out    Product between the matrix A and the Jacobian of the control with respect to the parameters
+   */
+  virtual void multiplyByDValue(double t, const Eigen::Ref<const VectorXs>& p, 
```
With the polynomial parametrizations the control is a linear function of the parameters `p`, so the derivatives do not depend on the values of `p`. However, in general the derivatives could depend on `p`, so we need to pass it.

andreadelprete commented on include/crocoddyl/core/integr-action-base.hpp:
```diff
@@ -0,0 +1,105 @@
+  using Base::u_ub_;                //!< Upper control limits
+  using Base::unone_;               //!< Neutral state
+
+  void init();
```
The reason is to avoid duplicating the code that initializes the object in the two constructors of the class. Actually I've seen this done many times and I don't see any downside.

cmastalli replied:
I understand the idea behind `init`, and we can keep this approach. However, you're still a bit duplicating the code as `IntegratedActionModelEulerTpl::init` implements the same code than the `IntegratedActionModelAbstractTpl::init()`.

 Note that when you construct your derived integrated-action, you call `Base(...)` which in turn calls `IntegratedActionModelAbstractTpl::init()`.

Similarly, in the `IntegratedActionModelRK4Tpl::init()`, there are bits that have been already implemented in the `IntegratedActionModelAbstractTpl::init()`. These lines of code needs to be removed: https://github.com/andreadelprete/crocoddyl/blob/polynomial-control/include/crocoddyl/core/integrator/rk4.hxx#L34-L45

andreadelprete commented on include/crocoddyl/core/controls/poly-one.hpp:
```diff
@@ -0,0 +1,65 @@
+ * The second half of p represents the value of u at time 0.5.
+ */
+template <typename _Scalar>
+class ControlPolyOneTpl : public ControlAbstractTpl<_Scalar> {
```
I am a bit short on time, so I'd rather skip this change as it seems less important than the others. We could always do this at a later time.

cmastalli replied:
Let's keep the current name as it follows a more clear standard, it is similar to other cases, e.g., `ControlPolyTwoRK4`.

I will mark this conversation as resolved!

andreadelprete commented on include/crocoddyl/core/controls/poly-two-rk4.hpp:
```diff
@@ -0,0 +1,68 @@
+ * because it requires the value of u exactly at 0, 0.5 and 1.
+ */
+template <typename _Scalar>
+class ControlPolyTwoRK4Tpl : public ControlAbstractTpl<_Scalar> {
```
The reason why `RK4` is in the name is that, all the possible parametrization of a 2nd-order polynomial, this is the one that is best suited to be used with an RK4 integration scheme because the three parameters represent exactly the value of u at time 0, 0.5 and 1, which are the times used by RK4 (or better, I should say the RK4 that is implemented in Crocoddyl). In another branch I've implemented also RK3, and another 2nd-order polynomial parametrization that is more suited for RK3, therefore I think it's important that RK4 remains in the name of this class. 

Regarding the idea of using "quadratic" instead of "poly2", the same comment I've left above applies here.

cmastalli replied:
> The reason why `RK4` is in the name is that, all the possible parametrization of a 2nd-order polynomial, this is the one that is best suited to be used with an RK4 integration scheme because the three parameters represent exactly the value of u at time 0, 0.5 and 1, which are the times used by RK4 (or better, I should say the RK4 that is implemented in Crocoddyl). In another branch I've implemented also RK3, and another 2nd-order polynomial parametrization that is more suited for RK3, therefore I think it's important that RK4 remains in the name of this class.

Thanks for the explanation. I have a follow-up question here. Can we use `ControlPolyTwoRK4` in other integrated-action models? If so, what are the cons of such choice? Or better could you elaborate a bit more the statement *"best suited to be used with an RK4"*.

If this control can only be used with the RK4 integrated-action model. Then, we should include a rule to prevent users to do that. I could provide you a way of doing it (if this needed).



> Regarding the idea of using "quadratic" instead of "poly2", the same comment I've left above applies here.

Let's keep `Poly2` as this naming choice seems to be more conventional in this case.

andreadelprete commented on include/crocoddyl/core/controls/poly-zero.hpp:
```diff
@@ -0,0 +1,61 @@
+ * @brief A polynomial function of time of degree zero, that is a constant
+ */
+template <typename _Scalar>
+class ControlPolyZeroTpl : public ControlAbstractTpl<_Scalar> {
```
Same comment as above.

cmastalli replied:
Marking this as resolved!

andreadelprete commented on bindings/python/crocoddyl/utils/quadruped.py:
```diff
@@ -68,7 +73,7 @@ def createCoMProblem(self, x0, comGoTo, timeStep, numKnots):
 
         # Defining the shooting problem
-        problem = crocoddyl.ShootingProblem(x0, comModels, comModels[-1])
+        problem = crocoddyl.ShootingProblem(x0, comModels[:-1], comModels[-1])
```
Couldn't do it today, but I'll do it asap

cmastalli replied:
I could help with this as well.
Just let me know if you're fine with it.

andreadelprete commented:
I think I've addressed most of the comments with the last commits. I've answered to some comments I disagreed with. The only thing I couldn't do yet is to check for the temporaries, but you can already start reviewing the code I think, considering I'll do that later.

cmastalli commented on bindings/python/crocoddyl/core/controls/poly-one.cpp:
```diff
@@ -0,0 +1,78 @@
+      .add_property("np", bp::make_function(&ControlParametrizationModelPolyOne::get_np),
+                    "dimension of the control parameters");
+}
+
```
We should create an specialized data structure and bind it here. See a more elaborated discussion, which motivates the idea of having specialized data structure, addressed in a comment below that I have partially copied here:

> The time-dependent coefficients of each control class should be stored in a specialized data, which in this case, the coefficients are .... Furthermore, I propose a standard name for these terms, ....
>
> **Why?**
> Crocoddyl API assumes that `calcDiff` is computed from any stored results in `calc`. Then, we are doing so if we stored this computed coefficient in `calc` and then reuse it in `calcDiff`. We can also introduce the same notion for `multiplyByJacobian` and `multiplyJacobianTransposeBy` ... .
>
> This also leads to the question: should we remove t from the `calc`, `calcDiff`, `multiplyByJacobian` and `multiplyJacobianTransposeBy`?. My personal view is a big YES as `t` can be seem as a model parameter rather than input. ...

Please do not response to this copied message here, use the discussion below :)

cmastalli replied:
I won't mark the need of creating and binding specialized data structures in the other cases. But be aware, that we need to do ti as well :)

cmastalli commented on bindings/python/crocoddyl/core/controls/poly-two-rk4.cpp:
```diff
@@ -0,0 +1,82 @@
+
+  bp::class_<ControlParametrizationModelPolyTwoRK4, bp::bases<ControlParametrizationModelAbstract> >(
+      "ControlParametrizationModelPolyTwoRK4",
+      "Constant control.\n\n"
+      "This control is a line function of time (normalized in [0,1])."
+      "The first half of the parameter vector contains the initial value of u, "
+      "whereas the second half contains the value of u at t=0.5.",
```
This comment has been copied from PolyOne class.
It needs to be updated.

cmastalli commented on bindings/python/crocoddyl/core/controls/poly-one.cpp:
```diff
@@ -0,0 +1,78 @@
+
+  bp::class_<ControlParametrizationModelPolyOne, bp::bases<ControlParametrizationModelAbstract> >(
+      "ControlParametrizationModelPolyOne",
+      "Constant control.\n\n"
```
Shouldn't be something like "Linear control"?

cmastalli commented on bindings/python/crocoddyl/core/controls/poly-zero.cpp:
```diff
@@ -0,0 +1,79 @@
+  bp::class_<ControlParametrizationModelPolyZero, bp::bases<ControlParametrizationModelAbstract> >(
+      "ControlParametrizationModelPolyZero",
+      "Constant control.\n\n"
+      "This control is a line function of time (normalized in [0,1])."
+      "The first half of the parameter vector contains the initial value of u, "
+      "whereas the second half contains the value of u at t=0.5.",
```
Same here

cmastalli commented on include/crocoddyl/core/control-base.hpp:
```diff
@@ -0,0 +1,187 @@
+   */
+  ControlParametrizationModelAbstractTpl(const std::size_t nu, const std::size_t np);
+  virtual ~ControlParametrizationModelAbstractTpl();
+
```
Cosmetic change that I will handle it (don't want to bother you @andreadelprete).

Let's put an order (an assert it in each respectively case) in the declaration. The order should be `calc`, `calcDiff`, `createData`, `checkData`, `convertBounds`, `multiplyByJacobian`, and `multiplyJacobianTransposeBy`.

cmastalli commented on include/crocoddyl/core/integrator/rk4.hpp:
```diff
@@ -123,24 +136,32 @@ struct IntegratedActionDataRK4Tpl : public ActionDataAbstractTpl<_Scalar> {
 
   VectorXs dx;
+  std::vector<VectorXs> u_diff;
   std::vector<boost::shared_ptr<DifferentialActionDataAbstractTpl<Scalar> > > differential;
```
Renamed this to differentials (a plural word). Note that `differential` is already stored in the abstract class.

cmastalli replied:
Both `controls` and `differentials` should be `std::array`, but this a tangent discussion. (please do no worry about it)

andreadelprete replied:
Ok for adding the "s", but note that `differential` is not stored in the abstract class.

cmastalli replied:
Then, I think we should include it in the base API. Do you have any point against to it?

Just let me know your feedback, I will handle it myself (as in any case I need to do some work on this class)

andreadelprete replied:
The reason why I didn't include differential in the base class is that in the case of Euler you just need one differential, whereas for RK4 you need 4 of them. We could of course have a single differential in the base class, and then add a 3d vector of differentials in the RK4 class, but that would break the current implementation of the RK4 integration, which uses a for loop assuming the 4 differential data structures are stored in a vector.

cmastalli replied:
I renamed back to differential, despite I liked more the plural one. The reason is to avoid deprecate the RK4 code.

cmastalli commented on include/crocoddyl/core/integrator/rk4.hpp:
```diff
@@ -58,26 +55,33 @@ class IntegratedActionModelRK4Tpl : public ActionModelAbstractTpl<_Scalar> {
   virtual void print(std::ostream& os) const;
```
We haven't implement this method (same for the Abstract and Euler integrators).

andreadelprete replied:
Probably I don't get your comment. The method seems to be implemented in the hxx file. Also for Euler.

cmastalli replied:
Yes, you're right. Apologies for the mistake.

Note:
It is still missing the abstract class, but this is a general ongoing issue in Crocoddyl. So, let's ignore it (so we can merge this work asap)

andreadelprete commented on include/crocoddyl/core/control-base.hpp:
```diff
@@ -0,0 +1,187 @@
+  }
+  virtual ~ControlParametrizationDataAbstractTpl() {}
+
+  VectorXs u_diff;    //!< value of the control
+  VectorXs u_params;  //!< value of the control parameters
```
Actually in that discussion, which involved also @proyan , we came to the conclusion of using the verbose names u_diff and u_params when referring to the vectors themselves, while using the shorter names w and u when referring to derivatives of quantities (such as cost or dynamics) with respect to u_diff and u_params.

cmastalli replied:
My understanding of the discussion is:
  1. @andreadelprete suggested to use a mix of short and long names in https://github.com/loco-3d/crocoddyl/pull/960#issuecomment-865679096
  2. @proyan didn't have any preference for `u_diff`, even he suggested what it could be considered a short name in https://github.com/loco-3d/crocoddyl/pull/960#issuecomment-866705880
  3. I proposed to use `w` as name in https://github.com/loco-3d/crocoddyl/pull/960#issuecomment-867039298, which it is indeed a short name as well. 

Unfortunately, mixing short and and long names as proposed in 1 is very confusing. Indeed, I struggled many times when I was reviewing the code. Examples of my struggles were: 
  - I wasn't clear what was the correct "control" to use in different parts of the code (but mainly in the integrators)
  - I confused many times if `w` (e.g., in `data->Lwu`) refer to the diff action or the control parametrization.

I would highly advice to do not follow the road proposed in 1. I really like much to have a single letter for everything. Otherwise, it starts to be a bit of *spaghetti code*. Reason why I would also highly advice to never break this rule of thumb.

proyan replied:
I agree with Andrea that using long descriptive names for vectors and shorter names for derivs would be easier. `u_diff` could explain more than `w` and demand less questions in the future.
But we are dragging out this point, I don't think it is that big an issue.

andreadelprete replied:
After going through the code for a while I agree with Carlos in the end. Having the two names is a bit confusing, so I've used `w` and `u` everywhere in the new version of the code.

andreadelprete commented on include/crocoddyl/core/control-base.hpp:
```diff
@@ -0,0 +1,187 @@
+
+  VectorXs u_diff;    //!< value of the control
+  VectorXs u_params;  //!< value of the control parameters
+  MatrixXs J;         //!< Jacobian of the control
```
Yes, the Jacobian at the moment is only used for debugging. Providing a default implementattion fo the other two metthods is a good idea, but not straightforward to implement because calcDiff takes the data as input, whereas these two metthods don't, so I should allocate a temporary Data structure inside the methhod, which seems a waste to me. Any idea for a workaround?

andreadelprete commented on unittest/test_controls.cpp:
```diff
@@ -0,0 +1,101 @@
+  // Computing
+  Eigen::MatrixXd A_J(Eigen::MatrixXd::Zero(A.rows(), control->get_np()));
+  Eigen::MatrixXd A_J_num_diff(Eigen::MatrixXd::Zero(A.rows(), control->get_np()));
+  control->multiplyByJacobian(t, p, A, A_J);
+  control_num_diff.multiplyByJacobian(t, p, A, A_J_num_diff);
```
Actually you don't need to compute the Jacobian to compute the product between Jacobian and a given matrix. that's hte reason why I've intrtoduced these methods.

andreadelprete commented on bindings/python/crocoddyl/core/control-base.cpp:
```diff
@@ -0,0 +1,75 @@
+           ":return p_lb, p_ub: lower and upper bounds on the control parameters (dim control.np).")
+      .def("multiplyByJacobian", pure_virtual(&ControlParametrizationModelAbstract_wrap::multiplyByJacobian_wrap),
+           bp::args("self", "t", "p", "A"),
+           "Compute the product between the given matrix A and the derivative of the control with respect to the "
+           "parameters.\n\n"
```
This is not true, so I won't add it.

andreadelprete commented on bindings/python/crocoddyl/core/control-base.cpp:
```diff
@@ -0,0 +1,75 @@
+          "multiplyJacobianTransposeBy",
+          pure_virtual(&ControlParametrizationModelAbstract_wrap::multiplyJacobianTransposeBy_wrap),
+          bp::args("self", "t", "p", "A"),
+          "Compute the product between the transpose of the derivative of the control with respect to the parameters\n"
+          "and a given matrix A.\n\n"
```
This is not true, so I won't add it.

andreadelprete commented on include/crocoddyl/core/integ-action-base.hpp:
```diff
@@ -0,0 +1,127 @@
+  /**
+   * @brief Return the dimension of the control input of the differential action model
+   */
+  std::size_t get_nu_diff() const;
```
In a previous comment you suggested to rename this method `get_nw`. Which of your comments should I follow?

andreadelprete commented on include/crocoddyl/core/integrator/euler.hpp:
```diff
@@ -58,39 +56,51 @@ class IntegratedActionModelEulerTpl : public ActionModelAbstractTpl<_Scalar> {
     const std::size_t ndx = model->get_state()->get_ndx();
+    const std::size_t nv = model->get_state()->get_nv();
     dx = VectorXs::Zero(ndx);
+    u_diff = VectorXs::Zero(model->get_nu_diff());
```
`get_nu` would not give me the correct size. I need the size of u_diff, not the size of u_params. I cannot directly access `nu_` even if it is protected because here I am not in the model class, but in the data class.

andreadelprete commented on include/crocoddyl/core/integ-action-base.hxx:
```diff
@@ -0,0 +1,111 @@
+template <typename Scalar>
+void IntegratedActionModelAbstractTpl<Scalar>::set_differential(
+    boost::shared_ptr<DifferentialActionModelAbstract> model) {
+  const std::size_t nu_diff = model->get_nu();
+  if (control_->get_nu() != nu_diff) {
+    control_->resize(nu_diff);
```
I think you got confused by the names. `control->get_nu` meant the size of the differential control, not the size of the control parameters. Anyway, following your suggestions, now `control->get_nu` will become `control->get_nw`, so everything should be clear.

andreadelprete commented on include/crocoddyl/core/integrator/euler.hpp:
```diff
@@ -58,39 +56,51 @@ class IntegratedActionModelEulerTpl : public ActionModelAbstractTpl<_Scalar> {
 
   boost::shared_ptr<DifferentialActionDataAbstractTpl<Scalar> > differential;
   VectorXs dx;
+  VectorXs u_diff;
```
I can remove it here, but in the data of RK4 u_diff is a vector of vectors, one for each internal step of RK4.

andreadelprete commented on include/crocoddyl/core/integrator/euler.hpp:
```diff
@@ -58,39 +56,51 @@ class IntegratedActionModelEulerTpl : public ActionModelAbstractTpl<_Scalar> {
   boost::shared_ptr<DifferentialActionDataAbstractTpl<Scalar> > differential;
   VectorXs dx;
+  VectorXs u_diff;
+  MatrixXs da_du;
```
`da_du` is used in `calcDiff` so I cannot remove it.

andreadelprete commented on include/crocoddyl/core/numdiff/control.hxx:
```diff
@@ -0,0 +1,96 @@
+void ControlParametrizationModelNumDiffTpl<Scalar>::calcDiff(
+    const boost::shared_ptr<ControlParametrizationDataAbstract>& data, double t,
+    const Eigen::Ref<const VectorXs>& p) const {
+  VectorXs tmp_p = VectorXs::Zero(np_);
+  VectorXs u0 = VectorXs::Zero(nu_);
```
Ok, I've added a Data structure for `ControlParametrizationModelNumDiff`, which contains `dp` and `u_diff0`.

andreadelprete commented on bindings/python/crocoddyl/core/control-base.cpp:
```diff
@@ -0,0 +1,75 @@
+          "and a given matrix A.\n\n"
+          ":param t: normalized time in [0, 1].\n"
+          ":param p: control parameters (dim control.np).\n"
+          ":param A: matrix to multiply (dim control.nu x na).\n"
```
I think there is no mistake in the documentation. Since `A` has to be multiplied on the left by the transposed Jacobian, and `J^T` has size npXnw, then `A` must have size nwXna, where na is arbitrary. In `multiplyByJacobian` instead, we compute the product `A*J`, so since `J` has size nwXnp, `A` must have size naXnw.

andreadelprete commented on include/crocoddyl/core/controls/poly-one.hxx:
```diff
@@ -0,0 +1,105 @@
+    throw_pretty("Invalid argument: "
+                 << "p has wrong dimension (it should be " + std::to_string(np_) + ")");
+  }
+  data->u_diff = (1 - 2 * t) * p.head(nu_) + 2 * t * p.tail(nu_);
```
At a first sight I don't like this suggestion, but let's talk about it. So far I've always seen the `data` object being used as an output data structure, where the methods `calc` and `calcDiff` would store their results. You're suggesting instead to add `t` inside `data`, which means that `data` would also be contains inputs to the methods. But this begs the question of why then `p` should remain in the API and not go inside `data` as well? 

Regarding the first suggestion of storing the coefficient of the polynomials inside data, I am not against it, but it seems a detail to me. It won't affect computation time significantly, and by itself it won't change the API, so it's really a minor change.

andreadelprete commented on include/crocoddyl/core/integrator/euler.hxx:
```diff
@@ -197,7 +182,10 @@ void IntegratedActionModelEulerTpl<Scalar>::quasiStatic(const boost::shared_ptr<
   boost::shared_ptr<Data> d = boost::static_pointer_cast<Data>(data);
 
-  differential_->quasiStatic(d->differential, u, x, maxiter, tol);
+  VectorXs uc(control_->get_nu());
+  differential_->quasiStatic(d->differential, uc, x, maxiter, tol);
+  control_->params(controlData_, 0.0, uc);
+  u = controlData_->u_params;
```
I cannot move the implementation in the parent class because the DifferentialActionData is introduced at the child class level. This is because in RK4 I need 4 instances of DifferentialActionData, so `d->differential` is a vector of DifferentialActionData in RK4, whereas it is a DifferentialActionData in Euler.

andreadelprete commented on include/crocoddyl/core/integrator/rk4.hpp:
```diff
@@ -94,24 +99,32 @@ struct IntegratedActionDataRK4Tpl : public ActionDataAbstractTpl<_Scalar> {
     }
 
     dx = VectorXs::Zero(ndx);
+    u_diff = std::vector<VectorXs>(4, VectorXs::Zero(nu_diff));
```
In Euler I could remove it because I could use the `w` variable of the controlParametrizationData. However, in RK4 I cannot do the same unless I create 4 instances of controlParametrizationData. At the moment I have only one instance of it.

andreadelprete commented on include/crocoddyl/core/integrator/rk4.hxx:
```diff
@@ -16,37 +16,51 @@ namespace crocoddyl {
+  if (time_step_ == Scalar(0.)) {
     enable_integration_ = false;
   }
   rk4_c_.push_back(Scalar(0.));
   rk4_c_.push_back(Scalar(0.5));
   rk4_c_.push_back(Scalar(0.5));
   rk4_c_.push_back(Scalar(1.));
```
This was not my choice, it was already there. Anyway we need the coefficients to be inside a vector for implementing the RK4 integration with a for loop.

cmastalli replied:
Ok, fair enough.

I won't ask you to fix previous code, this is my duty.
Let me handle it myself

andreadelprete commented on include/crocoddyl/core/integrator/rk4.hpp:
```diff
@@ -123,24 +136,32 @@ struct IntegratedActionDataRK4Tpl : public ActionDataAbstractTpl<_Scalar> {
   virtual ~IntegratedActionDataRK4Tpl() {}
 
   VectorXs dx;
+  std::vector<VectorXs> u_diff;
```
I've renamed `u_diff` as `ws`, but I prefer to maintain its size to 4 (instead of 3 as you suggest) to have a cleaner code in the RK4 integration loop. Otherwise I would have a shift in the indexes which would be hard to understand. The overhead in terms of memory is minimal, so I think it's worth it.

cmastalli replied:
Ok, it seems a reasonable design choice.

andreadelprete commented on bindings/python/crocoddyl/core/integ-action-base.hpp:
```diff
@@ -0,0 +1,67 @@
+///////////////////////////////////////////////////////////////////////////////
```
Since you offered help, I would gladly let you take care of the fixes to the Python bindings of IntegActionBase

andreadelprete commented on bindings/python/crocoddyl/core/integ-action-base.cpp:
```diff
@@ -0,0 +1,124 @@
+void exposeIntegratedActionAbstract() {
+  bp::register_ptr_to_python<boost::shared_ptr<IntegratedActionModelAbstract> >();
+
+  bp::class_<IntegratedActionModelAbstract_wrap, boost::noncopyable>(
```
I let you take care of this

andreadelprete commented on unittest/test_controls.cpp:
```diff
@@ -0,0 +1,101 @@
+  ts->add(BOOST_TEST_CASE(boost::bind(&test_multiplyJacobianTransposeBy_num_diff, control_type)));
+  framework::master_test_suite().add(ts);
+}
+
```
Yes we could. For instance, when using a constant control parametrization, all integration schemes should give the same result.

andreadelprete replied:
This however is something we could to when testing the integration schemes, not when testing the control parametrizations.

cmastalli replied:
As discussed, this might be still useful. Also to include warning messages to avoid an unreasonable user choice (i.e., higher controls order than the integrator).

Would you have the bandwidth for tackling these points?

andreadelprete replied:
Sure, I've just pushed the new unittests.

andreadelprete replied:
And now I've added a warning message when using a uselessly large control parametrization

andreadelprete commented:
@cmastalli As agreed, I stop working on the code now and I let you handle the last changes. Let me now if you need my help.

cmastalli reviewed (commented):
Hi @andreadelprete,

I have sorted out the remaining issues, with the exception of using substituting `std::vector` with `std::array` in the RK4 integrator. Unfortunately, I have a low bandwidth now, but I would try to provide a solution asap.

I know we have discussed quite a lot about the ControlParametrization API; however, I found it over complex when I was working in your PR.

The current design choice is *to follow calc, calcDiff, createData structure* implemented across many classes. However, it seems to me that this approach is complex as control classes are a bit more similar to `StateAbstract`. They do four things:
  1. Compute the control inputs `w` given `t` and a the control parameters `u` (defined in `calc`)
  2.  Define operators for Jacobian multiplications (defined in `multiplyByJacobian` and `multiplyJacobianTransposeBy`)
  3. Compute the control parameters `u` given `t` and the control inputs `w` (defined in `params`)
  4. Convert control input bounds into control parameters bounds (defined in `convertBounds`).

To simplify the API, I propose the following actions
  a. Renamed `calc` by `rollout` and include `w` as in/out argument
  b. Renamed `params` by `undo` and erase `t` from the arguments (I don't see why we need `t`).
  c. Removed `calcDiff` and `createData` as we don't use and need these methods, respectively.
 
The names on a and b are tentative, and I would like to discuss about it if you like my proposition.

My understanding is that @andreadelprete has started with a very different API, after @proyan suggested to implement through calc/calcDiff approach. I was in favor of @proyan's suggestions; however, as you see, I have changed my mind.

So, I guess @andreadelprete won't have any strong opinion here, but yet very welcome. About @proyan, could you rethink again your thought about this point?

cmastalli commented:
>  Regarding the name, I don't like `rollout` as it reminds me of the term used in Reinforcement Learning to mean "simulation". My previous choice for this method's name was "value".
> I like `undo` even less than `rollout`. My previous choice was `valueInv` (Inv stands for inverse). Indeed `t` is not needed in the current implementations, so we could remove it.

What about to keep `params` and rename `calc` to `inputs`?

andreadelprete commented:
> What about to keep `params` and rename `calc` to `inputs`?

My view is that the `ControlParametrizationModel` class basically implements a function w = f(u), mapping some input parameters `u` to some output values `w`. Inside the class there's nothing specific to the fact that it is used to represent the control inputs of a dynamical system. That's why I initially used some method names that were rather abstract, such as `value`, `Jacobian`, and `valueInv`. Switching to `calc`, `calcDiff` and `params` made sense because of consistency with the other classes of Crocoddyl. Now, if you wanna go back to an API similar to the initial one, I would use method names that are not specific to the use we make of this class because, who knows, one day this very same class could be used to parametrize something else besides the control trajectory (e.g., the state trajectory). For this reason, renaming `calc` to `inputs` does not make sense to me, because `calc` is the method that computes the function output `w = f(w)`. I would call it something like `compute`, or `value`, or `computeValue`, or `evaluate`.

proyan commented:
> My view is that the `ControlParametrizationModel` class basically implements a function w = f(u), mapping some input parameters `u` to some output values `w`

This is the reason that keeping `calc` `calcdiff` etc makes sense. Unlike in `StateAbstract`, which provides support for a specific state, there is a dynamic component `w=f(u)` to Control Parameters. It would be much easier to keep `calc` `calcDiff` as API, and maintain consistency, rather than asking the user to re-look the doc for this class.

cmastalli commented:
> This is the reason that keeping `calc` `calcdiff` etc makes sense. Unlike in `StateAbstract`, which provides support for a specific state, there is a dynamic component `w=f(u)` to Control Parameters. It would be much easier to keep `calc` `calcDiff` as API, and maintain consistency, rather than asking the user to re-look the doc for this class.

For accuracy, this is a mapping function rather than a "dynamic component". It maps u into w, furthermore there is no a state that evolves. Despite this technical inaccuracy, I have no concern with keeping the `calc` and `calcDiff` structure as @proyan substains. After all, the actuation models are also mapping functions (in this case depending on both state and control) and we follow the *calc&calcDiff* approach.

@andreadelprete and @proyan -- I am OK with keeping the current approach and I won't push for changing it anymore. However, @andreadelprete if you wish to push for a different API, then I would suggest to arrange a meeting (again my posture is neutral here).

Additionally, if we follow the current approach, then I have two more concerns (regarding the control-param API) that would like to broadly discuss first:
  1. `calcDiff` doesn't reuse computation of `calc`. This could be possible by including `t` as a data, but @andreadelprete disagreed on it.
  2. IMHO `params` should be embedded in `calc` function. In the end, `calc` functions need to use internally `u` (see these examples: https://github.com/andreadelprete/crocoddyl/blob/polynomial-control/include/crocoddyl/core/controls/poly-one.hxx#L26-L46, https://github.com/andreadelprete/crocoddyl/blob/polynomial-control/include/crocoddyl/core/controls/poly-two-rk4.hxx#L27-L61)

**Action points**
- @andreadelprete -- could you explain again your reasons for not going with 1. @proyan -- could you reply @andreadelprete on this topic?
- @andreadelprete -- could you let me know if you support 2.

andreadelprete commented:
> @andreadelprete and @proyan -- I am OK with keeping the current approach and I won't push for changing it anymore. However, @andreadelprete if you wish to push for a different API, then I would suggest to arrange a meeting (again my posture is neutral here).

I am fine with the current API.

>     1. `calcDiff` doesn't reuse computation of `calc`. This could be possible by including `t` as a data, but @andreadelprete disagreed on it.

You could do it. I don't see the benefit, that's all.

>     2. IMHO `params` should be embedded in `calc` function. In the end, `calc` functions need to use internally `u` (see these examples: https://github.com/andreadelprete/crocoddyl/blob/polynomial-control/include/crocoddyl/core/controls/poly-one.hxx#L26-L46, https://github.com/andreadelprete/crocoddyl/blob/polynomial-control/include/crocoddyl/core/controls/poly-two-rk4.hxx#L27-L61)

I don't see why you'd like to do that. `params` is used only for converting the bounds on `w` in bounds on `u`, and for converting the quasi-static `w` value in the corresponding `u` value. `calc` instead is used in very different locations of the code. Why do you wanna unify these two methods?

cmastalli reviewed (approved):
I am satisfied with the current state of the PR.
It is missing to replace std::vector by std::arrays; however, I am lacking of time for pushing this task forward (it requires some work on boost python).
